### DISCUSSION TITLE
Modification of Constraint/Motor After Creation

### DISCUSSION THREAD

**Discussion by PhoenixIllusion** [jrouwe/JoltPhysics.js#36] [Q&A] [answered]
I have been coding a BabylonJS Physics Plugin (v1 for now) using Jolt, and have most of the initial code there except some joint/constraint related issues.

I looked into the motor control, and was looking at how to increase or decrease the force. As a secondary plugin function, I also saw listed something like "change distance constraint". Based on my reading of the source code, it apperas the closest I could do for modifying any constraint values at all is call
```
   const newConstraintSettings = cast(currentConstraint.GetConstraintSettings(), NeededConstraintSettingCast)
   newConstraintSettings.motorSettings = XYZ;
   newConstraintSettings.<new settings> = XYZ
   physicsEngine.RemoveConstraint(currentConstraint);
   physicsEngine.AddConstraint(newConstraintSettings.Create(body1,body2))
``` 

I see some direct mappings of the Joints - https://doc.babylonjs.com/legacy/physics/joints
I know most of them can be worked with 6DOF if needed:

```
HingeConstraint  - Hinge - Hinge	Single Axis Rotation

* Need to determine if these are SwingTwistConstraint or ConeConstraint 
Ball and Socket - Ball and Socket	Multi Axis Rotation
Cone Twist - Unrestricted rotation (Twist) on One Axis, Limited Rotation (in Cone) On Others

Wheel - Two Axes Rotation

* Need to determine which of these is SliderConstraint
Slider - Single Axis Translation and Rotation
Prismatic - Prismatic	Single Axis Translation Only

DistanceConstraint - Distance - Set Bodies a Fixed Distance Apart
FixedConstraint - Locked - Bodies Act As One Body
```

I have skipped over soft-body physics for now, and did notice some awkward things (applying a point impulse to a center of mass still causes torque for some reason, CollisionGroup having void GetSubGroupID(), MeshShape needing to flip the triangle vertex order to collide vs render). I will need to add some more shapes ex: HeightField, ConvexHullShape.

**Comment by jrouwe** [comment]
> I looked into the motor control, and was looking at how to increase or decrease the force.

With motors you don't really set the force, you set the target velocity or position. You can limit the max force to apply though: MotorSettings::mMinForceLimit/mMaxForceLimit (which should usually be set to -MaxDesiredValue, MaxDesiredValue) or mMinTorqueLimit/mMaxTorqueLimit for rotational constraints.

> Based on my reading of the source code, it apperas the closest I could do for modifying any constraint values at all is call

I think I exposed all of the motor settings for all of the joints at run-time. E.g. the HingeConstraint::GetMotorSettings returns a reference to the motor settings that I think you can manipulate on the fly.

> I see some direct mappings of the Joints

I think these should be the mappings:

![image](https://github.com/jrouwe/JoltPhysics.js/assets/1621693/5b134b96-21ce-4ff5-a8c2-bee30f0c35d6)

> applying a point impulse to a center of mass still causes torque for some reason

That's quite odd, the responsible code is here:

https://github.com/jrouwe/JoltPhysics/blob/f8da353840a9a29b994439c81f2ddd0635062844/Jolt/Physics/Body/Body.inl#L155

and mPosition = COM, so the delta angular velocity should vanish if you apply an impulse at COM. Jolt is a bit odd w.r.t. COM, did you read [this](https://github.com/jrouwe/JoltPhysics/blob/master/Docs/Architecture.md#center-of-mass) part of the documentation?

> CollisionGroup having void GetSubGroupID()

Sorry about that, fixed now. Let me know if you need a new release.

> MeshShape needing to flip the triangle vertex order to collide vs render

That's odd, most engines I know use a counter clockwise triangle ordering. But if BabylonJS is clockwise then yes you'll have to flip.

**Comment by PhoenixIllusion** [comment]
I got a Hinge windmill setup, and tried to get a SliderConstraint setup.

My point-impulse issue was that I had tried to convert to local-space before applying, and that is already done inside the library so it wasn't actually applying to COM.

I got slider motor to work with a velocity I flip every so often (pictured) but am still trying to figure out how the slider position logic works. I was able to follow the Unit Test to jump the position once via the code, but after it landed at the target I wasn't able to force another target to drive towards via any positive or negative value.

``` 
  constraint.SetMotorState(EMotorState::Position);
  constraint.SetTargetPosition(Vec3(cMotorPos - cInitialPos).Dot(s.mSliderAxis1));
```
 
 
![2023-10-13-windmill-01](https://github.com/jrouwe/JoltPhysics.js/assets/756488/958d484e-1cd9-435b-ab42-21a495b9adca)

**Reply by PhoenixIllusion** [reply]
I have the example, without the velocity based bouncing, at my fork in regular Physics.JS:
https://github.com/PhoenixIllusion/JoltPhysics.js/blob/feature/motor-example/Examples/motor.html

Still can't get the SliderConstraint to move more than once with setPosition

**Reply by jrouwe** [reply, marked as answer]
Thanks for the demo! I've applied a few fixes and merged it into the mainline (together with your other fixes to the idl file).

> Still can't get the SliderConstraint to move more than once with setPosition

Your demo doesn't show what you're doing, but I'm guessing the object that's attached to the slider went to sleep because it reached the target and wasn't moving, so it doesn't respond to the new call to ```SetTargetPosition```. In Jolt you always have to explicitly wake an object using ```bodyInterface.ActivateBody(bodyId)``` (the only time Jolt does that for you is if an active object collides with a sleeping object).

**Reply by PhoenixIllusion** [reply]
That solved it. I can now make it jump around to anywhere I want on the line:
![2023-10-14-slider-position-01](https://github.com/jrouwe/JoltPhysics.js/assets/756488/4261c15e-74a8-4658-b544-c717c8eeee96)

**Reply by PhoenixIllusion** [reply]
![image](https://github.com/jrouwe/JoltPhysics.js/assets/756488/afbd6b18-3a29-449a-8468-4b8f80a43b17)

It took me a bit to figure out that the SetPosition meant:

Yellow - mSliderAxis
Blue Point - Target Body COM on axis
Red Point - mPoint2
Green Point - Box Body COM on axis
Magenta - example mPoint1 if you want to drive the surface of the box instead of COM. Above GIF sets mPoint1 to COM

Assuming the other point is on a static object:
Motor Position seems to work as "drive the Magenta Dot until it is Motor-Position-Value away from Red Dot along Yellow Line". Green dot (constrained dynamic object) will move relative to Magenta to enforce constraint.

**Comment by jrouwe** [comment]
That looks like a much better demo! Please send me a PR with your changes (in general please send me PR's for the demo's you're making, if I pull them myself then they get submitted (partially) under my name). 

W.r.t. what the position means: When configuring the joint you supply 2 points mPoint1 on body 1 and mPoint2 on body 2. This position by definition is position = 0. So if you set mPosition1 = (0 0 0) and mPosition2 = (1 0 0) and sliding axis (1 0 0) then that configuration will be position 0 and when body 2 moves to (0 0 0) position would be -1 because body 2 moved 1 unit against the sliding axis.
